src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -154,195 +155,260 @@ protected override void DetachCore()
+        /// </para>
+        /// <para>
+        /// If the peer is behind/equal to our best height an empty array is sent back.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessGetHeadersAsync(NetworkPeer peer, GetHeadersPayload getHeadersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(getHeadersPayload), getHeadersPayload);
 
-                        peer.SendMessageVoidAsync(headers);
-                        break;
-                    }
+            if (!this.CanRespondToGetHeaders)
+            {
+                this.logger.LogTrace("(-)[CANT_RESPOND_TO_HEADERS]");
+                return;
+            }
+
+            // Ignoring "getheaders" from peers because node is in initial block download unless the peer is whitelisted.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && !peer.Behavior<ConnectionManagerBehavior>().Whitelisted)
+            {
+                this.logger.LogTrace("(-)[IGNORE_ON_IBD]");
+                return;
+            }
+
+            HeadersPayload headers = new HeadersPayload();
+            ChainedBlock consensusTip = this.chainState.ConsensusTip;
+            consensusTip = this.Chain.GetBlock(consensusTip.HashBlock);
 
-                case HeadersPayload newHeaders:
+            ChainedBlock fork = this.Chain.FindFork(getHeadersPayload.BlockLocators);
+            if (fork != null)
+            {
+                if ((consensusTip == null) || (fork.Height > consensusTip.Height))
+                {
+                    // Fork not yet validated.
+                    fork = null;
+                }
+
+                if (fork != null)
+                {
+                    foreach (ChainedBlock header in this.Chain.EnumerateToTip(fork).Skip(1))
                     {
-                        // Represents the peers height from our point view.
-                        // This updates the pending tip parameter which is
-                        // the peers current best validated height.
-                        // If the peer's height is higher Chain.Tip is updated to have
-                        // the most PoW header.
-                        // It is sent in response to GetHeadersPayload or is solicited by the
-                        // peer when a new block is validated (and not in IBD).
-
-                        if (!this.CanSync)
+                        if (header.Height > consensusTip.Height)
                             break;
 
-                        if (newHeaders.Headers.Count == 0)
-                        {
-                            this.logger.LogTrace("Headers payload with 0 headers was received. Assume we're synced with the peer.");
+                        headers.Headers.Add(header.Header);
+                        if ((header.HashBlock == getHeadersPayload.HashStop) || (headers.Headers.Count == 2000))
```

Comment:
What is 2000? Consider using a constant or a comment to clarify what exactly that number represents.

Reply:
old code